### PR TITLE
Allow OOB locations

### PR DESCRIPTION
**Pull request by lionel-**
Adds a `oob = c("error", "extend")` parameter to `num_as_location()`. This is needed to implement things like `letters[1:30]`, e.g. in tibble.

The tricky part is to only allow consecutive extensions. From the output file:

```
can optionally extend beyond the end
====================================

> num_as_location(c(1:5, 7), 3, oob = "extend")
Error: Can't index beyond the end with non-consecutive locations.
i The input has size 3.
x The locations 4 and 7 are not consecutive.

> num_as_location(c(1:5, 7, 1), 3, oob = "extend")
Error: Can't index beyond the end with non-consecutive locations.
i The input has size 3.
x The locations 4 and 7 are not consecutive.

> num_as_location(c(1:5, 7, 1, 10), 3, oob = "extend")
Error: Can't index beyond the end with non-consecutive locations.
i The input has size 3.
x The locations 4, 7 and 10 are not consecutive.
```

### CODE REVIEW

File: R/subscript-loc.R [hunk 1]
```diff
@@ -58,23 +58,42 @@ vec_as_location <- function(i,
...
+    loc_oob = "error"
+  )
 }
 #' @rdname vec_as_location
 #' @param negative Whether to `"invert"` negative values to positive
 #'   locations, throw an informative `"error"`, or `"ignore"` them.
+#' @param oob If `"error"`, throws an informative `"error"` if some
+#'   elements are out-of-bounds. If `"extend"`, out-of-bounds
+#'   locations are allowed if they are consecutive after the end. This
+#'   can be used to implement extensible vectors like `letters[1:30]`.
```
Comment (DavisVaughan) [inline R/subscript-loc.R]: Is "extensible" right? "extendable" maybe? I found this a little confusing

File: tests/testthat/error/test-subscript-loc.txt [hunk 2]
```diff
@@ -272,3 +272,22 @@ x Can't subset element with unknown name `foo`.
...
+
+> num_as_location(c(1:5, 7, 1), 3, oob = "extend")
+Error: Can't index beyond the end with non-consecutive locations.
+i The input has size 3.
+x The locations 4 and 7 are not consecutive.
+
+> num_as_location(c(1:5, 7, 1, 10), 3, oob = "extend")
+Error: Can't index beyond the end with non-consecutive locations.
+i The input has size 3.
+x The locations 4, 7 and 10 are not consecutive.
```
Comment (hadley) [inline tests/testthat/error/test-subscript-loc.txt]: Should use Oxford comma: The locations 4, 7, and 10 are not consecutive

File: R/subscript-loc.R (hunk 1 above)
Comment (lionel-) [reply R/subscript-loc.R]: "stretchable" vectors?

File: tests/testthat/error/test-subscript-loc.txt (hunk 2 above)
Comment (lionel-) [reply tests/testthat/error/test-subscript-loc.txt]: Thanks, I'll fix this outside the PR because this will affect other tests.

**Pull request merged**